clap_complete/src/dynamic/completer.rs

```diff
@@ -290,15 +301,17 @@ fn complete_arg(
                 }
             }
 
-            if let Some(positional) = cmd
-                .get_positionals()
-                .find(|p| p.get_index() == Some(pos_index))
-            {
-                completions.extend(complete_arg_value(arg.to_value(), positional, current_dir));
-            }
+            if !takes_opt_flag {
+                if let Some(positional) = cmd
+                    .get_positionals()
+                    .find(|p| p.get_index() == Some(pos_index))
+                {
+                    completions.extend(complete_arg_value(arg.to_value(), positional, current_dir));
+                }
 
-            if let Ok(value) = arg.to_value() {
-                completions.extend(complete_subcommand(value, cmd));
+                if let Ok(value) = arg.to_value() {
+                    completions.extend(complete_subcommand(value, cmd));
+                }
```

Comment:
Whats with this change?  I'm not understanding the intent behind this and why `takes_opt_flag` is set in some cases but not others

Reply:
There is a case `--delimiter=comma,[TAB]`. If we do not distinguish between the value of the optional flag and the value of the positional argument, `--delimiter=comma` will be treated as the first value of the positional argument with `value_delimiter`, which can lead to unintended auto-completion.

> why `takes_opt_flag` is set in some cases but not others

If `arg` starts with `-` or `--`, it is treated as an optional flag. And if `arg.is_empty()` is true, we couldn't just treat it as an optional flag, so we don't set `takes_opt_flag`.

The above considerations do not account for the case where `allow_hyphen_values(true)` is enabled. There is the related issue https://github.com/clap-rs/clap/issues/5594

Reply:
I believe this affects existing behavior.  As a heads up, this is the type of change that is good to split into its own commit.

Overall, I feel like we are hitting the limits of what complexity we should have in this function.  There are a lot of conditional blocks in a row and how they all interact with each other is becoming increasingly unclear.  We need to look at how we can simplify the interactions to make the intent clearer.

Reply:
Yes, this does affect the existing behavior. Now our question can be reduced to whether it is desirable to complete the positional argument unconditionally (except for the index restriction) in the `ParseState::ValueDone` branch of the `complete_arg` function. We could discuss this as a separate issue, and in this PR I will remove `takes_opt_value`

Reply:
> Overall, I feel like we are hitting the limits of what complexity we should have in this function. There are a lot of conditional blocks in a row and how they all interact with each other is becoming increasingly unclear. We need to look at how we can simplify the interactions to make the intent clearer.

The code logic of the `ValueDone` branch of the `complete_arg` function is,
```
if arg.may_be_long_flag() || arg.may_be_long_flag(){
    // complete long flag
}
if arg.may_be_short_flag() || arg.may_be_long_flag(){
    // complete short flag
}
// complete positional argument

// complete subcommand
```
This results in possible duplication of `if` statements for each completion case.
When I added `takes_opt_flag`, I thought of a refactor code I wrote a long time ago. How about we do a refactor like this :
```
if arg.is_xx(){
    // complete subcommand, positional and optional flag, if any
} else if arg.is_xx(){
    // complete subcommand, positional and optional flag, if any
} else{
    // complete subcommand, positional and optional flag, if any
}
```
and reuse code blocks that generate completion results. This refactoring can make the if statement simpler, but there may be repeated completion function calls between two branches, such as `complete_subcommand`.

Reply:
Looks like this went away and so is no longer relevant